.pre-commit-config.yaml

```diff
@@ -0,0 +1,28 @@
+repos:
+  - repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v3.2.0
+    hooks:
+      - id: trailing-whitespace
+      - id: end-of-file-fixer
+      - id: check-added-large-files
+  - repo: https://github.com/scop/pre-commit-shfmt
+    rev: v3.7.0-1
+    hooks:
+      - id: shfmt # (requires/installs Go to build)
+  - repo: https://github.com/psf/black
+    rev: 22.10.0
+    hooks:
+      - id: black
```

Comment:
So I appreciate the spirit behind this PR and am fine with moving in this direction.  I already run black automatically on save so it doesn't make much difference to me either way.  However:

1.  We'd need to communicate the decision to use pre-commit because IIRC each user will still need to do some manual config to set it up locally.
2.  I'm not sure how this works on CI, so if a user skips that step they might end up contributing non-conformant code, and then we're back to whitespace whack-a-mole :)
3.  Doing a global reformatting will lose us a lot of `git blame` context, because this PR becomes the author of essentially the entire repo.  If we could configure this hook to work incrementally, only on changed files, we might have a better time of preserving history so people can tell who last touched a piece of code and when.
4. Another approach that might get us to the same desired end-state is to make black part of CI, that way a user can't get their pre-commit hooks misconfigured and gum up the repo.
5. We already have black rules stored in pyproject.toml (line-length, specifically).  I think the pre-commit hook automatically picks those up, but it's hard to tell here and would be worth having a comment that that was manually tested and works correctly.  If it doesn't and we have two sources of truth about black rules, it's going to be hard to sort out.